superpoint/evaluations/detector_evaluation.py

```diff
@@ -0,0 +1,222 @@
+    if simplified:
+        tp = np.any(matches, axis=1)  # keeps multiple matches for the same gt point
+        n_gt = np.sum(np.minimum(np.sum(matches, axis=0), 1))  # buggy
+    fp = np.logical_not(tp)
+    return tp, fp, prob, n_gt
+
+
+def div0(a, b):
+    with np.errstate(divide='ignore', invalid='ignore'):
+        c = np.true_divide(a, b)
+        idx = ~np.isfinite(c)
+        c[idx] = np.where(a[idx] == 0, 1, 0)  # -inf inf NaN
+    return c
+
+
+def compute_pr(exper_name, **kwargs):
+    """
+    Compute precision and recall.
+    """
+    # Gather TP and FP for all files
+    paths = get_paths(exper_name)
+    tp, fp, prob, n_gt = [], [], [], 0
+    for path in paths:
+        t, f, p, n = compute_tp_fp(np.load(path), **kwargs)
+        tp.append(t)
+        fp.append(f)
+        prob.append(p)
+        n_gt += n
+    tp = np.concatenate(tp)
+    fp = np.concatenate(fp)
+    prob = np.concatenate(prob)
+
+    # Sort in descending order of confidence
+    sort_idx = np.argsort(prob)[::-1]
+    tp = tp[sort_idx]
+    fp = fp[sort_idx]
+    prob = prob[sort_idx]
+
+    # Cumulative
+    tp_cum = np.cumsum(tp)
+    fp_cum = np.cumsum(fp)
+    recall = div0(tp_cum, n_gt)
+    precision = div0(tp_cum, tp_cum + fp_cum)
+    recall = np.concatenate([[0], recall, [1]])
+    precision = np.concatenate([[0], precision, [0]])
+    precision = np.maximum.accumulate(precision[::-1])[::-1]
+    return precision, recall, prob
+
+
+def compute_mAP(precision, recall):
+    """
+    Compute average precision.
+    """
+    return np.sum(precision[1:] * (recall[1:] - recall[:-1]))
+
+
+def compute_loc_error(exper_name, prob_thresh=0.5, distance_thresh=2):
+    """
+    Compute the localization error.
+    """
+    def loc_error_per_image(data):
+        # Read data
+        gt = np.where(data['keypoint_map'])
+        gt = np.stack([gt[0], gt[1]], axis=-1)
+        prob = data['prob']
+
+        # Filter out predictions
+        mask = np.where(prob > prob_thresh)
+        pred = np.array(mask).T
+        prob = prob[mask]
+
+        if not len(gt) or not len(pred):
+            return []
+
+        diff = np.expand_dims(pred, axis=1) - np.expand_dims(gt, axis=0)
+        dist = np.linalg.norm(diff, axis=-1)
+        dist = np.min(dist, axis=1)
+        correct_dist = dist[np.less_equal(dist, distance_thresh)]
+        return correct_dist
+    paths = get_paths(exper_name)
+    error = []
+    for path in paths:
+        error.append(loc_error_per_image(np.load(path)))
+    return np.mean(np.concatenate(error))
+
+
+def compute_repeatability(exper_name, prob_thresh=0.5, distance_thresh=3):
+    """
+    Compute the repeatability. The experiment must contain in its output the prediction
+    on 2 images, an original image and a warped version of it, plus the homography
+    linking the 2 images.
+    """
+    def warp_keypoints(keypoints, H):
+        warped_col0 = np.add(np.sum(np.multiply(keypoints, H[0, :2]), axis=1), H[0, 2])
```

Comment:
Would be faster to convert to homogeneous coordinates, matrix multiply, and convert back to heterogeneous (see TF equivalent [here](https://github.com/rpautrat/SuperPoint/blob/master/superpoint/models/utils.py#L146)).